src/tools/kdb/gen/elektra/elektragen.cpp

```diff
@@ -0,0 +1,439 @@
+{
+	std::string result;
+	result.resize (fileName.length ());
+	std::transform (fileName.begin (), fileName.end (), result.begin (), ::toupper);
+	escapeNonAlphaNum (result);
+	return result;
+}
+
+static inline std::string getArgName (const kdb::Key & key, kdb_long_long_t index, const std::string & defaultPrefix)
+{
+	auto indexStr = std::to_string (index);
+	auto metaName = "gen/arg/name/#" + std::string (indexStr.length () - 1, '_') + indexStr;
+	return key.hasMeta (metaName) ? key.getMeta<std::string> (metaName) : defaultPrefix + indexStr;
+}
+
+static inline std::string getArgDescription (const kdb::Key & key, kdb_long_long_t index, const std::string & kind)
+{
+	auto indexStr = std::to_string (index);
+	auto metaName = "gen/arg/description/#" + std::string (indexStr.length () - 1, '_') + indexStr;
+	return key.hasMeta (metaName) ? key.getMeta<std::string> (metaName) :
+					"Replaces occurence no. " + indexStr + " of " + kind + " in the keyname.";
+}
+
+static void getKeyArgs (const kdb::Key & key, const size_t parentKeyParts, kainjow::mustache::list & args, std::string & fmtString)
+{
+	using namespace kainjow::mustache;
+	auto parts = getKeyParts (key);
+	parts.erase (parts.begin (), parts.begin () + parentKeyParts);
+
+	std::stringstream fmt;
+
+	size_t pos = 1;
+	size_t names = 1;
+	size_t indices = 1;
+	for (const auto & part : parts)
+	{
+		if (part == "_")
+		{
+			const std::string & argName = getArgName (key, names, "name");
+			auto arg = object{ { "native_type", "const char *" },
+					   { "name", argName },
+					   { "code", argName },
+					   { "index?", false },
+					   { "description", getArgDescription (key, names, "_") } };
+			args.push_back (arg);
+			fmt << "%s/";
+			++pos;
+			++names;
+		}
+		else if (part == "#")
+		{
+			const std::string & argName = getArgName (key, indices, "index");
+
+			std::string argCode = "elektra_len (" + argName + "), elektra_len (";
+			argCode += argName + "), \"#___________________\", (long long) ";
+			argCode += argName;
+
+			auto arg = object{ { "native_type", "kdb_long_long_t" },
+					   { "name", argName },
+					   { "code", argCode },
+					   { "index?", true },
+					   { "description", getArgDescription (key, indices, "#") } };
+			args.push_back (arg);
+			fmt << "%*.*s%lld/";
+			pos += 4;
+			++indices;
+		}
+		else
+		{
+			// escape backslashes first too avoid collision
+			fmt << std::regex_replace (part, std::regex ("[\\\\/]"), "\\\\$0") << "/";
+		}
+	}
+
+	if (!args.empty ())
+	{
+		args.back ()["last?"] = true;
+	}
+
+	if (args.size () > 1)
+	{
+		args[args.size () - 2]["last_but_one?"] = true;
+	}
+
+	fmtString = fmt.str ();
+	fmtString.pop_back ();
+}
+
+static std::string keySetToCCode (kdb::KeySet set)
+{
+	using namespace kdb;
+	using namespace kdb::tools;
+
+	Modules modules;
+	PluginPtr plugin = modules.load ("c", KeySet ());
+
+	auto file = "/tmp/elektra.elektragen." + std::to_string (std::time (nullptr));
```

Comment:
This will be a problem if more than one `kdb gen` is invoked within one second. You could use mkdtemp if it is available?